### DISCUSSION TITLE
양방향 바인딩 vs 단방향 바인딩

### DISCUSSION THREAD

**Discussion by JUDONGHYEOK** [woowacourse-study/2022-woowahan-bansanghwe#111] [JavaScript]
## 바인딩이란?

 바인딩이라는 것은 말그대로 묶는 것을 의미한다. 두 데이터 혹은 정보의 소스들을 일치시키는 것을 의미한다. 데이터 바인딩이란 화면에 보이는 데이터와 브라우저 메모리에 있는 데이터를 일치 시키는 기법이다. 대다수의 자바스크립트 프레임워크가 단방향 데이터 바인딩을 지원하는 반면 Angular JS는 양방향 데이터 바인딩을 제공한다.

## 양방향 데이터 바인딩

데이터의 변화를 감지하여 템플릿과 결합해 화면을 갱신하여 화면 입력에 따라 데이터를 갱신하는 것이다. 데이터의 변경을 프레임 워크에서 감지하고 있다가 데이터가 변경되는 시점에 DOM 객체에 렌더링 해주거나 페이지 내에서 모델의 변경을 감지해 JS실행부에서 변경한다. 입력된 값이나 변경된 값에 따라 내용이 바로 바뀌기 때문에 따로 체크해주지 않아도 된다.

### 양방향 데이터의 장단점

장점

- 데이터가 자동으로 변경되기 때문에 코드 사용면에서 코드량을 줄여준다.

단점

- 변화에 따라 DOM 객체 전체를 렌더링 해주거나 데이터를 바꿔주므로 성능이 감소된다.
- 뷰에서도 데이터를 조작할 수 있고 데이터로서 뷰를 조작할 수도 있기 때문에 데이터 흐름을 예측하기 어렵다?

ex - MVC 패턴, Angluar, 

## 단방향 데이터 바인딩

 데이터와 템플릿을 결합해 화면을 생성한다. 사용자의 입력에 따라 데이터를 갱신하고 화면을 업데이트 해야하므로 단방향 데이터 바인딩으로 구성하면 화면을 업데이트 하는 코드를 매번 작성해주어야 한다. 

 리액트에서는 기본적으로 컴포넌트간의 데이터의 흐름은 최상위 부모 컴포넌트에서 차례로 내려오는 형식이다. 컴포넌트 내에서의 데이터 바인딩 방식은 state를 통해 이루어 지는데 state가 변경될 때는 setState를 통해 값을 바꿔주는 작업을 해야한다. input엘리먼트에서 change이벤트를 감지하여 state를 바꾸고 이를 통해 dom을 업데이트 하는 작업이 바로 단방향 바인딩이 되는 방식이다.

ex - flux패턴, React

### 단방향 데이터 바인딩의 장단점

장점

- 데이터의 흐름이 단방향이기 때문에 코드를 이해하기 쉽고 데이터의 추적과 디버깅이 쉽다.

단점

- 변화를 감지하고 화면을 업데이트하는 코드를 매번 작성해야 한다.

cf) - 뷰의 경우에는 컴포넌트간 통신은 단방향 바인딩을 사용하고 v-model을 통해 양방향 바인딩을 활용할 수 있다.

```tsx
new Vue({ 
  el: "#app", 
  data: {
    name: "Hoza"
  }
})
```

```html
<script src="https://cdn.jsdelivr.net/npm/vue"></script>
<div id="app">
  <input type="text" v-model="name">
  {{ name }}
</div>

```

cf)- 리액트도 초기에는 양방향 데이터 바인딩을 지원했다고 한다. 하지만 15버전부터 deprecated…

```jsx
var createReactClass = require('create-react-class');

var WithLink = createReactClass({
  mixins: [LinkedStateMixin],
  getInitialState: function() {
    return {message: 'Hello!'};
  },
  render: function() {
    return <input type="text" valueLink={this.linkState('message')} />;
  }
});
```

[https://reactjs.org/docs/two-way-binding-helpers.html](https://reactjs.org/docs/two-way-binding-helpers.html)

### 참고자료
[데이터 바인딩의 이해](https://sungjk.github.io/2015/11/22/AngularJS(2).html)

**Comment by hwangstar156** [comment]
바인딩에 대해서 알고 있는대로 말씀해주세요

**Comment by hwangstar156** [comment]
자바스크립트 바인딩에 대해서 설명해주세요

**Comment by hwangstar156** [comment]
call, apply의 차이점에 대해서 설명해주세요

**Comment by hwangstar156** [comment]
this에 대해서 설명해주세요

**Comment by hwangstar156** [comment]
this 바인딩의 종류에 대해서 설명해주세요

**Comment by hwangstar156** [comment]
함수, 메서드, 화살표 함수의 차이에 대해서 말해주세요

**Comment by hwangstar156** [comment]
화살표 함수에는 왜 this가 없을까요?

**Comment by hwangstar156** [comment]
양방향 데이터 바인딩과 단방향 데이터 바인딩에 대해서 설명해주세요

**Comment by hwangstar156** [comment]
단방향 데이터 바인딩을 사용하는 예시가 있나요?

**Comment by hwangstar156** [comment]
단방향 데이터 바인딩을 사용하며 생기는 문제점이 있나요?

**Comment by hwangstar156** [comment]
바닐라 js로만 단방향 데이터 바인딩을 구현하여 이용하신적이 있으신가요?

**Comment by hwangstar156** [comment]
제어 컴포넌트, 비제어 컴포넌트에 대해서 설명해주세요

**Comment by hwangstar156** [comment]
로그인 폼을 구현할떄 어떻게 이용하실껀가요?

**Comment by hwangstar156** [comment]
react는 왜 ref를 통한 비제어 컴포넌트를 지원하는 이유는 무엇인가요?

**Comment by hwangstar156** [comment]
mvc패턴을 이용했을때 장단점에 대해서 설명해주세요

**Comment by hwangstar156** [comment]
컴포넌트란?
